Create rake task for bulk reloading of SuggestedResource records

matt-bernhardt opened the pull request:
## Developer

This defines a rake task for importing SuggestedResource records from a CSV file, which must be loaded from a remote URL (via either HTTP or HTTPS protocols, to support local environments). The rake task hands off to a new `bulk_replace` method on the Detector::SuggestedResource class, which actually does the deletion and re-import steps.

There are a few side effects to this change, which are discussed in more detail in the commit message.

Of particular note is the drop in code coverage, which seems to happen because this is the first PR to add tests for any rake task? The task defined here is being reported as 100% coverage - but I'm not sure what our practices are with rake testing in general.

### Ticket(s)

https://mitlibraries.atlassian.net/browse/TCO-52

### Accessibility

- [ ] ANDI or Wave has been run in accordance to [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened 
      as new issues (link to those issues in the Pull Request details above)
- [x] There are no accessibility implications to this change

### Documentation

- [ ] Project documentation has been updated
- [x] No documentation changes are needed

### ENV

- [ ] All new ENV is documented in README.
- [ ] All new ENV has been added to Heroku Pipeline, Staging and Prod.
- [x] ENV has not changed.

### Stakeholders

- [ ] Stakeholder approval has been confirmed
- [x] Stakeholder approval is not needed

### Dependencies and migrations

NO dependencies are updated
NO migrations are included

## Reviewer

### Code

- [ ] I have confirmed that the code works as intended.
- [ ] Any CodeClimate issues have been fixed or confirmed as
added technical debt.

### Documentation

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message).
- [ ] The documentation has been updated or is unnecessary.
- [ ] New dependencies are appropriate or there were no changes.

### Testing

- [ ] There are appropriate tests covering any new functionality.
- [ ] No additional test coverage is required.

jazairi commented on app/models/detector/suggested_resource.rb:
```diff
@@ -53,5 +53,27 @@ def calculate_fingerprint(old_phrase)
+      raise ArgumentError.new, 'Tabular CSV is required' unless input.instance_of?(CSV::Table)
+
+      # Need to check what columns exist in input
+      required_headers = %w[Title URL Phrase]
+      missing_headers = required_headers - input.headers
```
It seems like the inputs are case-sensitive, in which case I'd suggest we underscore that in the method documentation above. (The exact values are already listed, but it may be helpful to explicitly note case sensitivity.)

matt-bernhardt replied:
Yeah, good point - case sensitivity did bite me initially with this (the File path silently downcases the header values, while URI doesn't (among other changes). I'll add this to the docs.

jazairi commented on lib/tasks/suggested_resources.rake:
```diff
@@ -0,0 +1,28 @@
+
+    url = URI.parse(args.addr)
+
+    raise ArgumentError.new, 'HTTP/HTTPS scheme is required' unless url.scheme.in?(%w[http https])
```
Does it make sense to only accept HTTPS? (I'd hope that any CSV we're loading is coming through HTTPS and not HTTP, but the Internet is weird sometimes.)

jazairi replied:
Oh, good call. I wasn't thinking about local. I don't see it as a huge issue, fwiw -- no need to make a change.

jazairi reviewed (changes requested):
Nice work on this! I have a couple of questions and suggestions to improve the documentation, but overall this is looking great. I especially appreciate the robust test coverage for the rake task. I think we normally test those indirectly, but reviewing this made me reconsider whether that's the right approach. I agree with your suggestion to make that a topic for team discussion.

Speaking of which, I should note that I haven't confirmed functionality locally. The coverage is so thorough that I'm not too worried about it. I can do that before approving, though, if you think it would be prudent.

matt-bernhardt commented on test/tasks/suggested_resource_rake_test.rb:
```diff
@@ -0,0 +1,68 @@
+    records_before = Detector::SuggestedResource.count # We have three fixtures at the moment
+    first_record_before = Detector::SuggestedResource.first
+    VCR.use_cassette('suggested_resource:reload from remote csv') do
+      remote_file = 'http://static.lndo.site/suggested_resources.csv'
```
No problem - I can add that.

matt-bernhardt commented:
Okay, @jazairi - I've added a note to the parameter for the rake task to clarify that the column names are case sensitive, and also added a section to the readme for regenerating the cassettes (including the `.lando.yml` file I'm using to serve these files locally).

jazairi reviewed (approved):
Looks great -- thanks!
